chefkoch/recipe.py

```diff
@@ -0,0 +1,771 @@
+            for node in self.nodes:
+                nodeIsValid = True
+                for key in node.inputs:
+                    input = node.inputs[key]
+                    # to do: Werte direkt zulassen, nicht nur ?ber flavour
+                    # WARN schmei?en
+                    # python logs verwenden: kein flavour, kein output, also
+                    # interpretiert als string
+                    # chef analyse (from log)
+                    inputIsValid = False
+                    prefix = os.path.splitext(input)[0]
+                    extension = os.path.splitext(input)[1]
+                    if (input in outputs_of_all_nodes or
+                            prefix == 'flavour' or
+                            extension == '.json'):
+                        inputIsValid = True
+                    else:
+                        if (os.path.isfile(input)):
+                            inputIsValid = True
+                    if not inputIsValid:
+                        nodeIsValid = False
+                if not nodeIsValid:
+                    unreachable_nodes.append(node)
+
+        # 3. Delete unreachable nodes and unreachable outputs and do it again.
+            try_again = (len(unreachable_nodes)>0)
+            for node in unreachable_nodes:
+                warn = (("" if warn is None else warn) + 'Node ' + node.name +
+                        ' or one of its previous nodes has an invalid input' +
+                        ' and therefore cannot be computed. ')
+                for key in node.outputs:
+                    output = node.outputs[key]
+                    outputs_of_all_nodes.remove(output)
+
+                # outputs_of_all_nodes.remove(node.outputs.values())
+                # # <-- if remove can remove a list of objects
+                # keys_to_remove = [val for val in node.outputs.values()
+                # if val in output_of_all_nodes]
+                # outputs_of_all_nodes = [oo for oo in outputs_of_all_nodes
+                # if oo not in node.outputs.values()]
+                self.nodes.remove(node)
+
+        # 4. Loop until all nodes are reachable
+        return err, warn
+
+    def findCircles(self):
+        """
+        Makes list of all nodes, that have only flavour parameters as inputs.
+        Then starts depth-first search for every root node. If there is a way
+        back to a previously visited node, there is a warning about a circle.
+        
+        :returns:
+            err - String about circle or "" if everything is correct
+        """
+        rootNodes = []
+        # 1. Make list of all nodes that only have flavour inputs.
+        for node in self.nodes:
+            isRootNode = True
+            for key in node.inputs:
+                input = node.inputs[key]
+                if not input.startswith("flavour."):
+                    isRootNode = False
+                    break
+            if isRootNode:
+                rootNodes.append(node)
```

Comment:
Same here as above: Maybe there are codings that avoid single-element `append` calls or `in` decisions by _vectorizing_ it a bit. This often improves readability, but mostly increases speed for larger sets (already noticeable at a few hundred elements)

Reply:
Vielleicht fällt dir etwas ein. :blush:

```rootNodes = []
        # 1. Make list of all root nodes
        # 1.1. Make set of all inputs that refere to the flavour file
        all_ins = set(n.inputs.values() for n in self.nodes)
        flavour_inputs = [i for i in all_ins if i.startswith("flavour.")]
        # TODO neue Möglichkeit für startswith auf Mengen finden
        # 1.2. If all inputs of a node are in there, node is root node.
        for node in self.nodes:
            node_inputs = set(node.inputs.values())
            if (node_inputs & flavour_inputs) == node_inputs:
                # in that case all of node's inputs start with flavour.
                rootNodes.append(node)
```

Der Teil `i.startswith("flavour.")` funktioniert nicht, da die Elemente keine strings sind, sondern dict_values und ich weiß nicht wie ich sie auspacke ohne doch wieder langsame Einzeloperationen zu verwenden

```all_ins = set([]) 
        for node in self.nodes:
            all_ins.update(node.inputs.values())
```
Ich hab es jetzt einfach so gemacht, vektorisierter wird es leider nicht :confused: